apps/cmp.c

```diff
@@ -0,0 +1,3531 @@
+        } else if (opt_tls_cert == NULL) {
+            CMP_err("missing -tls_cert option");
+        }
+    }
+
+    if (opt_msg_timeout >= 0) /* must do this before setup_ssl_ctx() */
+        (void)OSSL_CMP_CTX_set_option(ctx, OSSL_CMP_OPT_MSG_TIMEOUT,
+                                      opt_msg_timeout);
+    if (opt_total_timeout >= 0)
+        (void)OSSL_CMP_CTX_set_option(ctx, OSSL_CMP_OPT_TOTAL_TIMEOUT,
+                                      opt_total_timeout);
+
+    if (opt_reqin != NULL || opt_reqout != NULL
+            || opt_rspin != NULL || opt_rspout != NULL || opt_use_mock_srv)
+        (void)OSSL_CMP_CTX_set_transfer_cb(ctx, read_write_req_resp);
+
+    if (opt_tls_used) {
+        APP_HTTP_TLS_INFO *info = OPENSSL_zalloc(sizeof(*info));
+
+        if (opt_use_mock_srv) {
+            CMP_err("cannot use TLS options together with -use_mock_srv");
+            goto err;
+        }
+        if (info == NULL)
+            goto err;
+        (void)OSSL_CMP_CTX_set_http_cb_arg(ctx, info);
+        info->server = opt_server;
+        info->port = server_port_s;
+        info->use_proxy = opt_proxy != NULL;
+        info->timeout = OSSL_CMP_CTX_get_option(ctx, OSSL_CMP_OPT_MSG_TIMEOUT);
+        info->ssl_ctx = setup_ssl_ctx(ctx, e);
+        if (info->ssl_ctx == NULL)
+            goto err; /* info will be freed along with CMP ctx */
+#ifndef OPENSSL_NO_SOCK
+        (void)OSSL_CMP_CTX_set_http_cb(ctx, app_http_tls_cb);
+#endif
+    }
+
+    if (!setup_protection_ctx(ctx, e))
+        goto err;
+
+    if (!setup_request_ctx(ctx, e))
+        goto err;
+
+    if (!set_name(opt_recipient, OSSL_CMP_CTX_set1_recipient, ctx, "recipient")
+            || !set_name(opt_expect_sender, OSSL_CMP_CTX_set1_expected_sender,
+                         ctx, "expected sender"))
+        goto oom;
+
+    if (opt_geninfo != NULL && !handle_opt_geninfo(ctx))
+        goto err;
+
+    ret = 1;
+
+ err:
+    OPENSSL_free(proxy_host);
+    OPENSSL_free(proxy_port_str);
+    return ret;
+ oom:
+    CMP_err("out of memory");
+    goto err;
+}
+
+/* Very basic HTTP server: initializiation */
+static BIO *init_http_server(const char *port)
+{
+#ifdef OPENSSL_NO_SOCK
+    CMP_err("Error setting up accept BIO - sockets not supported");
+    return NULL;
+#else
+    BIO *acbio = NULL, *bufbio;
+
+    bufbio = BIO_new(BIO_f_buffer());
+    if (bufbio == NULL)
+        goto err;
+    acbio = BIO_new(BIO_s_accept());
+    if (acbio == NULL
+        || BIO_set_bind_mode(acbio, BIO_BIND_REUSEADDR) < 0
+        || BIO_set_accept_port(acbio, port) < 0) {
+        CMP_err("Error setting up accepting BIO");
+        goto err;
+    }
+
+    BIO_set_accept_bios(acbio, bufbio);
+    bufbio = NULL;
+    if (BIO_do_accept(acbio) <= 0) {
+        CMP_err("Error starting accepting");
+        goto err;
+    }
+
+    return acbio;
+
+ err:
+    BIO_free_all(acbio);
+    BIO_free(bufbio);
+    return NULL;
+#endif
+}
+
+/* Very basic HTTP server: read request */
```

Comment:
I'll try tomorrow...

Reply:
@mattcaswell, reusing HTTP server code between `apps/{ocsp.c,cmp.c}` was certainly a good thought.
So I've extracted the HTTP server code into a new `apps/lib/http_server.c`, adding also `apps/include/http_server.h`. This turned out to be much more work than expected. 
On the bright side this reduced much redundancy, and also any future HTTP server functionality in the apps can take advantage of that.
On this occasion I fixed a couple of memory leaks and diagnostic messages in `http_server_get_req()` and some issues in case `OPENSSL_NO_SOCK`.

Reply:
Since a lot of code has been moved, should we keep this major cleanup as part of this PR or convert it as a separate PR?
In the latter case I'd have to introduce a couple of stubs in `apps/cmp.c` such that this PR still builds while that separate PR is not yet merged.

Reply:
Since this PR is already quite enormous, I'd prefer to split it out into a separate PR.

Reply:
All right, done in the new PR #11736.

This carve-out removes nearly 100 lines from `apps/cmp.c`.
I've added two TODOs how to re-adapt the code when #11736 is merged.